Features

AhmadKHamdan opened the pull request.

AhmadKHamdan committed:
Implemented the first phase of Flight class and has been tested in the main

AhmadKHamdan committed:
Seperated Flight classes in Flight directory. Flight class is DTO. FlightValidator is for validation. FlightExtensions is for extension methods. Enum FlightClass has been seperated in its own directory. Valdiate price method is done.

AhmadKHamdan committed:
Add flight data import functionality and validation methods

- Introduced FlightImporter class to handle flight data import from CSV.
- Added methods for reading and parsing flight data.
- Implemented validation methods in FlightValidator for various flight attributes.
- Updated Flight class constructor to utilize validation methods.
- Enhanced FlightExtensions with a method to print all flights.

AhmadKHamdan committed:
- Implement flight validation logic and improve error handling during flight import and print detailed list of errors

- Fix variable naming inconsistency in FlightImporter class

AhmadKHamdan committed:
Add Booking and BookingExtensions classes for managing flight bookings and calculating prices

AhmadKHamdan committed:
Refactor Booking properties to use getters and setters; enhance BookingExtensions with booking management methods and detailed string representation; update Program to demonstrate booking functionality.

AhmadKHamdan committed:
Refactor Booking class to use integer BookingId; enhance BookingExtensions with flight search and booking functionality; add FlightSearchService for filtering flights based on user criteria; update Program for improved user interaction and menu navigation.

AhmadKHamdan committed:
Refactored namespaces and imports to follow standards; Implemented booking filtering for managers; Moved the handlers for search criteria to a separate class; Refactored the criteria display to include current criteria; Added flight imports to the manager menu

AhmadKHamdan committed:
Added "Current" to the booking criteria menu print method

AhmadKHamdan committed:
Implemented dynamic model validation details;  Added attributes to the flight class fields; Added checks for empty flights or bookings lists with feedback

oalataiwi commented on Booking/BookingExtensions.cs:
```diff
@@ -0,0 +1,292 @@
+                }
+            }
+        }
+    }
```
do early return plz

oalataiwi commented on Booking/BookingExtensions.cs:
```diff
@@ -0,0 +1,292 @@
+        }
+    }
+
+    public static void SearchAndBook(this List<Booking> bookings, List<Flight> flights)
```
this seems to break SRP, it performs 2 things : search and book!

oalataiwi commented on Booking/BookingExtensions.cs:
```diff
@@ -0,0 +1,292 @@
+                for (int i = 0; i < filteredFlights.Count; i++)
+                {
+                    Console.WriteLine($"{i + 1}. {filteredFlights[i].ToDetailedString()}");
+                }
```
why not using foreach?!

AhmadKHamdan replied:
I need to print the number of each booking so the user can enter the number of the booking he desires

oalataiwi commented on Flight/FlightValidator.cs:
```diff
@@ -0,0 +1,247 @@
+        if (string.IsNullOrWhiteSpace(flightNumber))
+            throw new ArgumentException("Flight number cannot be empty.");
+        if (flightNumber.Length < 3 || flightNumber.Length > 6)
+            throw new ArgumentException("Flight number must be between 3 and 10 characters.");
```
these should be placed in one place and be referenced wherever u want them

AhmadKHamdan replied:
I did not understand what you meant

oalataiwi replied:
i mean the hardcoded error msgs

oalataiwi commented on Booking/BookingSearchService.cs:
```diff
@@ -0,0 +1,159 @@
+            return $"≤ {maxPrice:C}";
+
+        return value.ToString()!;
+    }
```
utility can be moved to another class

oalataiwi commented on Booking/BookingSearchService.cs:
```diff
@@ -0,0 +1,159 @@
+
+public class BookingSearchService
+{
+    private readonly object[] UserCriteria = new object[8];
```
can be done as SearchObject and define all attributes as nullable
and pass that object to search method, and perform checks based on non-nullable fields

oalataiwi commented:
plz add entityRepo layer to deal with bookings, flights via file system storage
and move any business related logic to service layer 
the only way to modify or access the data is the repo layer

AhmadKHamdan committed:
Refactor flight management system:
- Implemented FlightsRepository for flight data handling.
- Removed  FlightImporter.cs as data is now managed through FlightRepository.
- Updated FlightExtensions to remove unused methods.
- Created FlightSearchService to utilize FlightSearchCriteria for filtering.
- Introduced BookingService and BookingsRepository for managing bookings.
- Added BookingSearchCriteria for improved booking search functionality.
- Updated Program.cs to integrate new services and streamline flight and booking management.
- Minor changes to improve readability and functionality

oalataiwi commented on Booking/Booking.cs:
```diff
@@ -0,0 +1,38 @@
+
+    public Booking(int bookingId, Flight flight, int passengerId, FlightClass flightClass, DateTime bookingDate, int numberOfSeats)
+    {
+        BookingId = bookingId;
```
what if this id is taken? i think u should be consistent in the way u populate ids

AhmadKHamdan replied:
As long as no one manually edits the data (csv files), the ID will be consistent because it will start from the highest ID in the file. I didn't want to do something more strict for the ID because in a normal day, I will have a database and not a file.
But if you want, I can fix it

oalataiwi replied:
here u assign the id without checking if it is > nextBookingId

oalataiwi commented on Booking/BookingSearchCriteria.cs:
```diff
@@ -0,0 +1,15 @@
+
+namespace Airport_Ticket_Booking_System.Bookings;
+
+public class BookingSearchCriteria
```
NIT: BookingSearchQuery

oalataiwi commented on Handlers/SearchCriteriaHandlers.cs:
```diff
@@ -0,0 +1,144 @@
+    {
+        while (true)
+        {
+            Console.Write("Enter max price: ");
```
why max?

AhmadKHamdan replied:
I didn't think it made sense to search for flights that cost exactly 200 for example
So I went for max price instead

oalataiwi commented on Booking/BookingSearchService.cs:
```diff
@@ -0,0 +1,160 @@
+
+            var filteredBookings = FilterBookings(bookings);
+
+            if (filteredBookings.Count == 0)
+            {
+                Console.Write("\nNo matching bookings. Retry filtering? (y/n): ");
+
+                if (Console.ReadLine()?.ToLower() != "y")
+                {
+                    return new List<Booking>();
+                }
+
+                continue;
```
i think we can just return the empty list, no retry is needed here

oalataiwi commented on Booking/BookingSearchService.cs:
```diff
@@ -0,0 +1,160 @@
+        return query.ToList();
+    }
+
+    private void ShowCriteriaMenu()
```
i think ShowCriteriaMenu and HandleCriteriaMenuInput can be moved to another utility service

oalataiwi commented on Booking/BookingsRepository.cs:
```diff
@@ -0,0 +1,91 @@
+                var columns = line.Split(',');
+                try
+                {
+                    var bookingId = int.Parse(columns[0]);
```
use tryparse instead on parse

oalataiwi commented on Flight/Flight.cs:
```diff
@@ -0,0 +1,72 @@
+        int bookedSeats
+    )
+    {
+        FlightNumber = FlightValidator.ValidateFlightNumber(flightNumber);
+        DepartureCountry = FlightValidator.ValidateCountry(departureCountry);
+        DepartureAirport = FlightValidator.ValidateAirport(departureAirport);
+        DepartureDateTime = FlightValidator.ValidateDepartureDateTime(departureDateTime);
+        DestinationCountry = FlightValidator.ValidateCountry(destinationCountry);
+        DestinationAirport = FlightValidator.ValidateAirport(destinationAirport);
+        FlightDuration = FlightValidator.ValidateFlightDuration(flightDuration);
+        Capacity = FlightValidator.ValidateCapacity(capacity);
+        BookedSeats = FlightValidator.ValidateBookedSeats(bookedSeats, capacity);
+        PricePerClass = FlightValidator.ValidatePricePerClass(pricePerClass);
```
validation should not live inside the model. create a separate validator that knows how to validate a Flight (or a Flight request), and call that validator before you add the object to your collection.

oalataiwi commented on Flight/FlightSearchService.cs:
```diff
@@ -0,0 +1,141 @@
+
+namespace Airport_Ticket_Booking_System.Flights;
+
+public class FlightSearchService
```
same comments as BookingSearchService

oalataiwi commented on Flight/FlightService.cs:
```diff
@@ -0,0 +1,36 @@
+
+    public FlightService()
+    {
+        _flights = FlightsRepository.GetFlights();
```
the service should nott maintain its own list of flights. managing the collection of flights is the repo's responsibility. the service’s role is simply to call repository methods that operate on the repo’s internal flight list

oalataiwi commented on Flight/FlightsRepository.cs:
```diff
@@ -0,0 +1,128 @@
+                    var capacity = int.Parse(columns[8]);
+                    var bookedSeats = int.Parse(columns[9]);
+
+                    var errors = FlightValidator.GetValidationErrors(
+                        flightNumber,
+                        departureCountry,
+                        departureAirport,
+                        departureDateTime,
+                        destinationCountry,
+                        destinationAirport,
+                        flightDuration,
+                        prices,
+                        capacity,
+                        bookedSeats
+                    );
+
+                    if (errors.Count > 0)
+                    {
+                        Console.WriteLine($"Validation errors for flight {flightNumber}: {string.Join("; ", errors)}");
+                        continue;
+                    }
+
+                    var flight = new Flight(
+                        flightNumber,
+                        departureCountry,
+                        departureAirport,
+                        departureDateTime,
+                        destinationCountry,
+                        destinationAirport,
+                        flightDuration,
+                        prices,
+                        capacity,
+                        bookedSeats
+                    );
```
Instead of putting validation inside the Flight model, you can centralize it in a validator. For example:

var flight = new Flight(...);
var errors = FlightValidator.GetValidationErrors(flight);

if (errors.Count > 0)
{
    // handle errors (e.g. show messages, throw, etc.)
}
else
{
    flights.Add(flight);
}


This way, the model stays lean, and the validator is responsible for enforcing the business rules before the flight gets added.

AhmadKHamdan committed:
Rename BookingSearchCriteria to BookingSearchQuery; Remove retry filtering option

AhmadKHamdan committed:
Refactor booking and flight search services to utilize CriteriaUtilityService; Renamed FlightSearchCriteria to FlightSearchQuery

AhmadKHamdan committed:
Fixed an error with filtering books as it was not used in the program

AhmadKHamdan committed:
Used TryParse instead of Parse; Introduced ValidationMessages class for consistent messaging.

AhmadKHamdan committed:
Implement FlightFactory for flight creation and validation; refactor Flight and FlightsRepository to utilize new factory method

AhmadKHamdan committed:
Refactor BookingService to eliminate private bookings list; Retrieve bookings directly from repository; Same thing with flights; Changed flights validator to take a flight object rather than the flight's information; Used TryParse instead of Parse in flight repository

AhmadKHamdan committed:
Fixed the issue with making sure booking ID is sequential and unique
